// @sentry/remix/esm/utils/instrumentServer.js
import { _optionalChain } from '@sentry/utils';
import { getActiveTransaction, hasTracingEnabled, runWithAsyncContext } from '@sentry/core';
import { captureException, getCurrentHub } from '@sentry/node';
import { logger, addExceptionMechanism, fill, loadModule, tracingContextFromHeaders, isNodeEnv, dynamicSamplingContextToSentryBaggageHeader } from '@sentry/utils';
import { getFutureFlagsServer, getRemixVersionFromBuild } from './futureFlags.js';
import { isResponse, extractData, isRouteErrorResponse, isDeferredData, json, matchServerRoutes, getRequestMatch } from './vendor/response.js';
import { normalizeRemixRequest } from './web-fetch.js';

/* eslint-disable max-lines */

let FUTURE_FLAGS;
let IS_REMIX_V2;

// Flag to track if the core request handler is instrumented.
let isRequestHandlerWrapped = false;

const redirectStatusCodes = new Set([301, 302, 303, 307, 308]);
function isRedirectResponse(response) {
  return redirectStatusCodes.has(response.status);
}

function isCatchResponse(response) {
  return response.headers.get('X-Remix-Catch') != null;
}

async function extractResponseError(response) {
  const responseData = await extractData(response);

  if (typeof responseData === 'string') {
    return responseData;
  }

  if (response.statusText) {
    return response.statusText;
  }

  return responseData;
}

/**
 * Captures an exception happened in the Remix server.
 *
 * @param err The error to capture.
 * @param name The name of the origin function.
 * @param request The request object.
 *
 * @returns A promise that resolves when the exception is captured.
 */
async function captureRemixServerException(err, name, request) {
  // Skip capturing if the thrown error is not a 5xx response
  // https://remix.run/docs/en/v1/api/conventions#throwing-responses-in-loaders
  if (IS_REMIX_V2) {
    if (isRouteErrorResponse(err) && err.status < 500) {
      return;
    }
  } else if (isResponse(err) && err.status < 500) {
    return;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let normalizedRequest = request ;

  try {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    normalizedRequest = normalizeRemixRequest(request );
  } catch (e) {
    (typeof __SENTRY_DEBUG__ === 'undefined' || __SENTRY_DEBUG__) && logger.warn('Failed to normalize Remix request');
  }

  captureException(isResponse(err) ? await extractResponseError(err) : err, scope => {
    const activeTransactionName = _optionalChain([getActiveTransaction, 'call', _ => _(), 'optionalAccess', _2 => _2.name]);

    scope.setSDKProcessingMetadata({
      request: {
        ...normalizedRequest,
        // When `route` is not defined, `RequestData` integration uses the full URL
        route: activeTransactionName
          ? {
              path: activeTransactionName,
            }
          : undefined,
      },
    });

    scope.addEventProcessor(event => {
      addExceptionMechanism(event, {
        type: 'instrument',
        handled: false,
        data: {
          function: name,
        },
      });

      return event;
    });

    return scope;
  });
}

function makeWrappedDocumentRequestFunction(remixVersion) {
  return function (origDocumentRequestFunction) {
    return async function (

      request,
      responseStatusCode,
      responseHeaders,
      context,
      loadContext,
    ) {
      let res;

      const activeTransaction = getActiveTransaction();

      try {
        const span = _optionalChain([activeTransaction, 'optionalAccess', _3 => _3.startChild, 'call', _4 => _4({
          op: 'function.remix.document_request',
          origin: 'auto.function.remix',
          description: activeTransaction.name,
          tags: {
            method: request.method,
            url: request.url,
          },
        })]);

        res = await origDocumentRequestFunction.call(
          this,
          request,
          responseStatusCode,
          responseHeaders,
          context,
          loadContext,
        );

        _optionalChain([span, 'optionalAccess', _5 => _5.finish, 'call', _6 => _6()]);
      } catch (err) {
        if (!_optionalChain([FUTURE_FLAGS, 'optionalAccess', _7 => _7.v2_errorBoundary]) && remixVersion !== 2) {
          await captureRemixServerException(err, 'documentRequest', request);
        }

        throw err;
      }

      return res;
    };
  };
}

function makeWrappedDataFunction(
  origFn,
  id,
  name,
  remixVersion,
) {
  return async function ( args) {
    let res;
    const activeTransaction = getActiveTransaction();
    const currentScope = getCurrentHub().getScope();

    try {
      const span = _optionalChain([activeTransaction, 'optionalAccess', _8 => _8.startChild, 'call', _9 => _9({
        op: `function.remix.${name}`,
        origin: 'auto.ui.remix',
        description: id,
        tags: {
          name,
        },
      })]);

      if (span) {
        // Assign data function to hub to be able to see `db` transactions (if any) as children.
        currentScope.setSpan(span);
      }

      res = await origFn.call(this, args);

      currentScope.setSpan(activeTransaction);
      _optionalChain([span, 'optionalAccess', _10 => _10.finish, 'call', _11 => _11()]);
    } catch (err) {
      if (!_optionalChain([FUTURE_FLAGS, 'optionalAccess', _12 => _12.v2_errorBoundary]) && remixVersion !== 2) {
        await captureRemixServerException(err, name, args.request);
      }

      throw err;
    }

    return res;
  };
}

const makeWrappedAction =
  (id, remixVersion) =>
  (origAction) => {
    return makeWrappedDataFunction(origAction, id, 'action', remixVersion);
  };

const makeWrappedLoader =
  (id, remixVersion) =>
  (origLoader) => {
    return makeWrappedDataFunction(origLoader, id, 'loader', remixVersion);
  };

function getTraceAndBaggage() {
  const transaction = getActiveTransaction();
  const currentScope = getCurrentHub().getScope();

  if (isNodeEnv() && hasTracingEnabled()) {
    const span = currentScope.getSpan();

    if (span && transaction) {
      const dynamicSamplingContext = transaction.getDynamicSamplingContext();

      return {
        sentryTrace: span.toTraceparent(),
        sentryBaggage: dynamicSamplingContextToSentryBaggageHeader(dynamicSamplingContext),
      };
    }
  }

  return {};
}

function makeWrappedRootLoader(remixVersion) {
  return function (origLoader) {
    return async function ( args) {
      const res = await origLoader.call(this, args);
      const traceAndBaggage = getTraceAndBaggage();

      if (isDeferredData(res)) {
        res.data['sentryTrace'] = traceAndBaggage.sentryTrace;
        res.data['sentryBaggage'] = traceAndBaggage.sentryBaggage;
        res.data['remixVersion'] = remixVersion;

        return res;
      }

      if (isResponse(res)) {
        // Note: `redirect` and `catch` responses do not have bodies to extract.
        // We skip injection of trace and baggage in those cases.
        // For `redirect`, a valid internal redirection target will have the trace and baggage injected.
        if (isRedirectResponse(res) || isCatchResponse(res)) {
          (typeof __SENTRY_DEBUG__ === 'undefined' || __SENTRY_DEBUG__) &&
            logger.warn('Skipping injection of trace and baggage as the response does not have a body');
          return res;
        } else {
          const data = await extractData(res);

          if (typeof data === 'object') {
            return json(
              { ...data, ...traceAndBaggage, remixVersion },
              { headers: res.headers, statusText: res.statusText, status: res.status },
            );
          } else {
            (typeof __SENTRY_DEBUG__ === 'undefined' || __SENTRY_DEBUG__) &&
              logger.warn('Skipping injection of trace and baggage as the response body is not an object');
            return res;
          }
        }
      }

      return { ...res, ...traceAndBaggage, remixVersion };
    };
  };
}

/**
 * Creates routes from the server route manifest
 *
 * @param manifest
 * @param parentId
 */
function createRoutes(manifest, parentId) {
  return Object.entries(manifest)
    .filter(([, route]) => route.parentId === parentId)
    .map(([id, route]) => ({
      ...route,
      children: createRoutes(manifest, id),
    }));
}

/**
 * Starts a new transaction for the given request to be used by different `RequestHandler` wrappers.
 *
 * @param request
 * @param routes
 * @param pkg
 */
function startRequestHandlerTransaction(
  hub,
  name,
  source,
  request

,
) {
  const { traceparentData, dynamicSamplingContext, propagationContext } = tracingContextFromHeaders(
    request.headers['sentry-trace'],
    request.headers.baggage,
  );
  hub.getScope().setPropagationContext(propagationContext);

  const transaction = hub.startTransaction({
    name,
    op: 'http.server',
    origin: 'auto.http.remix',
    tags: {
      method: request.method,
    },
    ...traceparentData,
    metadata: {
      source,
      dynamicSamplingContext: traceparentData && !dynamicSamplingContext ? {} : dynamicSamplingContext,
    },
  });

  hub.getScope().setSpan(transaction);
  return transaction;
}

/**
 * Get transaction name from routes and url
 */
function getTransactionName(
  routes,
  url,
  pkg,
) {
  const matches = matchServerRoutes(routes, url.pathname, pkg);
  const match = matches && getRequestMatch(url, matches);
  return match === null ? [url.pathname, 'url'] : [match.route.id, 'route'];
}

function wrapRequestHandler(origRequestHandler, build) {
  const routes = createRoutes(build.routes);
  const pkg = loadModule('react-router-dom');

  return async function ( request, loadContext) {
    return runWithAsyncContext(async () => {
      const hub = getCurrentHub();
      const options = _optionalChain([hub, 'access', _13 => _13.getClient, 'call', _14 => _14(), 'optionalAccess', _15 => _15.getOptions, 'call', _16 => _16()]);
      const scope = hub.getScope();

      let normalizedRequest = request;

      try {
        normalizedRequest = normalizeRemixRequest(request);
      } catch (e) {
        (typeof __SENTRY_DEBUG__ === 'undefined' || __SENTRY_DEBUG__) && logger.warn('Failed to normalize Remix request');
      }

      const url = new URL(request.url);
      const [name, source] = getTransactionName(routes, url, pkg);

      scope.setSDKProcessingMetadata({
        request: {
          ...normalizedRequest,
          route: {
            path: name,
          },
        },
      });

      if (!options || !hasTracingEnabled(options)) {
        return origRequestHandler.call(this, request, loadContext);
      }

      const transaction = startRequestHandlerTransaction(hub, name, source, {
        headers: {
          'sentry-trace': request.headers.get('sentry-trace') || '',
          baggage: request.headers.get('baggage') || '',
        },
        method: request.method,
      });

      const res = (await origRequestHandler.call(this, request, loadContext)) ;

      if (isResponse(res)) {
        transaction.setHttpStatus(res.status);
      }

      transaction.finish();

      return res;
    });
  };
}

/**
 * Instruments `remix` ServerBuild for performance tracing and error tracking.
 */
function instrumentBuild(build) {
  const routes = {};

  const remixVersion = getRemixVersionFromBuild(build);
  IS_REMIX_V2 = remixVersion === 2;

  const wrappedEntry = { ...build.entry, module: { ...build.entry.module } };

  // Not keeping boolean flags like it's done for `requestHandler` functions,
  // Because the build can change between build and runtime.
  // So if there is a new `loader` or`action` or `documentRequest` after build.
  // We should be able to wrap them, as they may not be wrapped before.
  const defaultExport = wrappedEntry.module.default ;
  if (defaultExport && !defaultExport.__sentry_original__) {
    fill(wrappedEntry.module, 'default', makeWrappedDocumentRequestFunction(remixVersion));
  }

  for (const [id, route] of Object.entries(build.routes)) {
    const wrappedRoute = { ...route, module: { ...route.module } };

    const routeAction = wrappedRoute.module.action ;
    if (routeAction && !routeAction.__sentry_original__) {
      fill(wrappedRoute.module, 'action', makeWrappedAction(id, remixVersion));
    }

    const routeLoader = wrappedRoute.module.loader ;
    if (routeLoader && !routeLoader.__sentry_original__) {
      fill(wrappedRoute.module, 'loader', makeWrappedLoader(id, remixVersion));
    }

    // Entry module should have a loader function to provide `sentry-trace` and `baggage`
    // They will be available for the root `meta` function as `data.sentryTrace` and `data.sentryBaggage`
    if (!wrappedRoute.parentId) {
      if (!wrappedRoute.module.loader) {
        wrappedRoute.module.loader = () => ({});
      }

      // We want to wrap the root loader regardless of whether it's already wrapped before.
      fill(wrappedRoute.module, 'loader', makeWrappedRootLoader(remixVersion));
    }

    routes[id] = wrappedRoute;
  }

  return { ...build, routes, entry: wrappedEntry };
}

function makeWrappedCreateRequestHandler(
  origCreateRequestHandler,
) {
  // To track if this wrapper has been applied, before other wrappers.
  // Can't track `__sentry_original__` because it's not the same function as the potentially manually wrapped one.
  isRequestHandlerWrapped = true;

  return function ( build, ...args) {
    FUTURE_FLAGS = getFutureFlagsServer(build);
    const newBuild = instrumentBuild(build);
    const requestHandler = origCreateRequestHandler.call(this, newBuild, ...args);

    return wrapRequestHandler(requestHandler, newBuild);
  };
}

/**
 * Monkey-patch Remix's `createRequestHandler` from `@remix-run/server-runtime`
 * which Remix Adapters (https://remix.run/docs/en/v1/api/remix) use underneath.
 */
function instrumentServer() {
  const pkg = loadModule('@remix-run/server-runtime');

  if (!pkg) {
    (typeof __SENTRY_DEBUG__ === 'undefined' || __SENTRY_DEBUG__) && logger.warn('Remix SDK was unable to require `@remix-run/server-runtime` package.');

    return;
  }

  fill(pkg, 'createRequestHandler', makeWrappedCreateRequestHandler);
}

export { captureRemixServerException, createRoutes, getTransactionName, instrumentBuild, instrumentServer, isRequestHandlerWrapped, startRequestHandlerTransaction };
