// msw/lib/core/handlers/GraphQLHandler.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};
var GraphQLHandler_exports = {};
__export(GraphQLHandler_exports, {
  GraphQLHandler: () => GraphQLHandler,
  isDocumentNode: () => isDocumentNode
});
module.exports = __toCommonJS(GraphQLHandler_exports);
var import_RequestHandler = require("./RequestHandler.js");
var import_getTimestamp = require("../utils/logging/getTimestamp.js");
var import_getStatusCodeColor = require("../utils/logging/getStatusCodeColor.js");
var import_serializeRequest = require("../utils/logging/serializeRequest.js");
var import_serializeResponse = require("../utils/logging/serializeResponse.js");
var import_matchRequestUrl = require("../utils/matching/matchRequestUrl.js");
var import_parseGraphQLRequest = require("../utils/internal/parseGraphQLRequest.js");
var import_getPublicUrlFromRequest = require("../utils/request/getPublicUrlFromRequest.js");
var import_devUtils = require("../utils/internal/devUtils.js");
var import_getRequestCookies = require("../utils/request/getRequestCookies.js");
function isDocumentNode(value) {
  if (value == null) {
    return false;
  }
  return typeof value === "object" && "kind" in value && "definitions" in value;
}
class GraphQLHandler extends import_RequestHandler.RequestHandler {
  constructor(operationType, operationName, endpoint, resolver, options) {
    let resolvedOperationName = operationName;
    if (isDocumentNode(operationName)) {
      const parsedNode = (0, import_parseGraphQLRequest.parseDocumentNode)(operationName);
      if (parsedNode.operationType !== operationType) {
        throw new Error(
          `Failed to create a GraphQL handler: provided a DocumentNode with a mismatched operation type (expected "${operationType}", but got "${parsedNode.operationType}").`
        );
      }
      if (!parsedNode.operationName) {
        throw new Error(
          `Failed to create a GraphQL handler: provided a DocumentNode with no operation name.`
        );
      }
      resolvedOperationName = parsedNode.operationName;
    }
    const header = operationType === "all" ? `${operationType} (origin: ${endpoint.toString()})` : `${operationType} ${resolvedOperationName} (origin: ${endpoint.toString()})`;
    super({
      info: {
        header,
        operationType,
        operationName: resolvedOperationName
      },
      resolver,
      options
    });
    this.endpoint = endpoint;
  }
  parse(args) {
    return __async(this, null, function* () {
      const parsedResult = yield (0, import_parseGraphQLRequest.parseGraphQLRequest)(args.request).catch(
        (error) => {
          console.error(error);
          return void 0;
        }
      );
      if (typeof parsedResult === "undefined") {
        return void 0;
      }
      return {
        query: parsedResult.query,
        operationType: parsedResult.operationType,
        operationName: parsedResult.operationName,
        variables: parsedResult.variables
      };
    });
  }
  predicate(args) {
    if (!args.parsedResult) {
      return false;
    }
    if (!args.parsedResult.operationName && this.info.operationType !== "all") {
      const publicUrl = (0, import_getPublicUrlFromRequest.getPublicUrlFromRequest)(args.request);
      import_devUtils.devUtils.warn(`Failed to intercept a GraphQL request at "${args.request.method} ${publicUrl}": anonymous GraphQL operations are not supported.

Consider naming this operation or using "graphql.operation()" request handler to intercept GraphQL requests regardless of their operation name/type. Read more: https://mswjs.io/docs/api/graphql/#graphqloperationresolver`);
      return false;
    }
    const hasMatchingUrl = (0, import_matchRequestUrl.matchRequestUrl)(
      new URL(args.request.url),
      this.endpoint
    );
    const hasMatchingOperationType = this.info.operationType === "all" || args.parsedResult.operationType === this.info.operationType;
    const hasMatchingOperationName = this.info.operationName instanceof RegExp ? this.info.operationName.test(args.parsedResult.operationName || "") : args.parsedResult.operationName === this.info.operationName;
    return hasMatchingUrl.matches && hasMatchingOperationType && hasMatchingOperationName;
  }
  extendResolverArgs(args) {
    var _a, _b, _c;
    const cookies = (0, import_getRequestCookies.getAllRequestCookies)(args.request);
    return {
      query: ((_a = args.parsedResult) == null ? void 0 : _a.query) || "",
      operationName: ((_b = args.parsedResult) == null ? void 0 : _b.operationName) || "",
      variables: ((_c = args.parsedResult) == null ? void 0 : _c.variables) || {},
      cookies
    };
  }
  log(args) {
    return __async(this, null, function* () {
      var _a, _b, _c, _d;
      const loggedRequest = yield (0, import_serializeRequest.serializeRequest)(args.request);
      const loggedResponse = yield (0, import_serializeResponse.serializeResponse)(args.response);
      const statusColor = (0, import_getStatusCodeColor.getStatusCodeColor)(loggedResponse.status);
      const requestInfo = ((_a = args.parsedResult) == null ? void 0 : _a.operationName) ? `${(_b = args.parsedResult) == null ? void 0 : _b.operationType} ${(_c = args.parsedResult) == null ? void 0 : _c.operationName}` : `anonymous ${(_d = args.parsedResult) == null ? void 0 : _d.operationType}`;
      console.groupCollapsed(
        import_devUtils.devUtils.formatMessage(
          `${(0, import_getTimestamp.getTimestamp)()} ${requestInfo} (%c${loggedResponse.status} ${loggedResponse.statusText}%c)`
        ),
        `color:${statusColor}`,
        "color:inherit"
      );
      console.log("Request:", loggedRequest);
      console.log("Handler:", this);
      console.log("Response:", loggedResponse);
      console.groupEnd();
    });
  }
}
